File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function (nums, target) {
    return [binarySearch(nums, target, true), binarySearch(nums, target, false)]
};


const binarySearch = (nums, target, isFirstSearch) => {
    let left = 0;
    let right = nums.length - 1;
    let ans = -1;
    while (left <= right) {
        let mid = left + Math.floor((right - left) / 2);
        if (nums[mid] > target) {
            right = mid - 1;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            ans = mid;
            if (isFirstSearch) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
    }

    return ans;
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        result = [-1, -1]
        
        result[0] = self.binarySeach(nums, target, True)
        result[1] = self.binarySeach(nums, target, False)
        
        return result
    
    def binarySeach(self, nums, target, isLeftSearching):
        index = -1
        low, high = 0, len(nums) -1
        
        while low <= high:
            mid = low + (high - low)//2
            if nums[mid] == target:
                index = mid
                if isLeftSearching:
                    high = mid - 1
                else:
                    low = mid + 1
            elif nums[mid] > target: 
                high = mid - 1
            else:
                 low = mid + 1
            
        return index
        